### PR TITLE
Added supported versions changes to bot client

### PR DESCRIPTION
**Pull request by bharatsuri97**

### CODE REVIEW

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 1]
```diff
@@ -86,11 +96,34 @@ void setHealthApi(HealthApi healthApi) {
...
       BotSendMessageRequest botSendMessageRequest) {
 
+    String currentApiVersion;
+    try {
+      currentApiVersion = getCurrentApiVersion();
+      SupportedVersions versions = getSupportedVersions();
+      List<SupportedVersionsVersions> supportedVersions = versions.getVersions()
+          .stream()
+          .filter(v -> {
+            assert v.getVersionNumber() != null;
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:116]: Assertions can be disabled using `-disableassertions` jvm flag. It's not recommended to use assertions.

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 2]
```diff
@@ -86,11 +96,34 @@ void setHealthApi(HealthApi healthApi) {
...
+            return v.getVersionNumber().equals(currentApiVersion);
+          })
+          .collect(Collectors.toList());
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:119]: I don't think you need to collect to list and check size. Instead, can we find first match by using `findFirst()` method in the stream?

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 3]
```diff
@@ -221,6 +254,16 @@ public Status getHealthStatus() {
     }
   }
 
+  public SupportedVersions getSupportedVersions() {
+    CompletableFuture<SupportedVersions> versionsFuture = versionsApi.versionsGet().toFuture();
+
+    try {
+      return versionsFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:263]: Can we add an exception message ?

File: src/test/java/com/salesforce/einsteinbot/sdk/client/ClientApiWireMockTest.java [hunk 4]
```diff
@@ -309,4 +335,15 @@ private void stubStatusResponse(String responseBodyFile) {
     );
   }
 
+  private void stubVersionsResponse(String responseBodyFile) {
```
Comment (relango) [inline src/test/java/com/salesforce/einsteinbot/sdk/client/ClientApiWireMockTest.java:338]: Can we add tests for the case when Versions endpoint returns ErrorResponse?

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 5]
```diff
@@ -86,11 +96,34 @@ void setHealthApi(HealthApi healthApi) {
...
+  void setVersionsApi(VersionsApi versionsApi) {
+    this.versionsApi = versionsApi;
+  }
+
   @Override
   public BotResponse startChatSession(RequestConfig config,
       ExternalSessionId sessionId,
       BotSendMessageRequest botSendMessageRequest) {
 
+    String currentApiVersion;
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:109]: Can we move this logic to a separate method ?

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 6]
```diff
@@ -86,11 +96,34 @@ void setHealthApi(HealthApi healthApi) {
...
+      SupportedVersions versions = getSupportedVersions();
+      List<SupportedVersionsVersions> supportedVersions = versions.getVersions()
+          .stream()
+          .filter(v -> {
+            assert v.getVersionNumber() != null;
+            return v.getVersionNumber().equals(currentApiVersion);
+          })
+          .collect(Collectors.toList());
+      if (supportedVersions.size() != 1) {
+        throw new RuntimeException("Unable to start chat. API version not supported.");
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:121]: Do we want to do this check only for `startChatsession` or to all other methods like `sendMessage` and `endSession` ?

**Review by relango — CHANGES_REQUESTED** [review]

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 1 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:116]: Refactored this to not have `assert`

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 2 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:119]: Yep, makes sense. Updated this

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 3 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:263]: Added an exception message

File: src/test/java/com/salesforce/einsteinbot/sdk/client/ClientApiWireMockTest.java (hunk 4 above)
Comment (bharatsuri97) [reply src/test/java/com/salesforce/einsteinbot/sdk/client/ClientApiWireMockTest.java:338]: Added this

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 6 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:121]: @atulkc What are your thoughts on this? My thought process was that if the check is done when session is starting, then that is enough

File: pom.xml [hunk 7]
```diff
@@ -260,6 +260,22 @@
...
+                <artifactId>properties-maven-plugin</artifactId>
+                <version>1.0.0</version>
+                <executions>
+                    <execution>
+                        <phase>generate-resources</phase>
+                        <goals>
+                            <goal>write-project-properties</goal>
+                        </goals>
+                        <configuration>
+                            <outputFile>${project.build.directory}/classes/properties-from-pom.properties</outputFile>
```
Comment (relango) [inline pom.xml:274]: How is this used?

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 5 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:109]: Yep, refactoring this

**Review by relango — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 7 above)
Comment (bharatsuri97) [reply pom.xml:274]: In order to know in code, what is the api version, I copied over the pom properties including the property `api-spec-yaml-file` using this plugin, and then read the property in code.

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 8]
```diff
@@ -246,4 +282,23 @@ private Mono<ClientResponse> mapErrorResponse(ClientResponse clientResponse) {
...
   }
+
+  private String getCurrentApiVersion() throws IOException {
+    Properties properties;
+    InputStream is = getClass().getClassLoader()
+        .getResourceAsStream("properties-from-pom.properties");
+    properties = new Properties();
+    properties.load(is);
+    return properties.getProperty("api-spec-yaml-file").substring(1, 6).replace("_", ".");
+  }
```
Comment (bharatsuri97) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:293]: @relango  The maven plugin will copy over the pom properties into this file that is read here to know which api version is being used. I could not find a way to read it as it is hardcoded in the auto-gen code.

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 9]
```diff
@@ -246,4 +282,23 @@ private Mono<ClientResponse> mapErrorResponse(ClientResponse clientResponse) {
             .error(new ChatbotResponseException(clientResponse.statusCode(), errorDetails,
                 clientResponse.headers())));
   }
+
+  private String getCurrentApiVersion() throws IOException {
+    Properties properties;
+    InputStream is = getClass().getClassLoader()
+        .getResourceAsStream("properties-from-pom.properties");
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:289]: Can you share the example content of `properties-from-pom.properties`?

**Review by relango — CHANGES_REQUESTED** [review]

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 9 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:289]:
```
#Properties
#Tue Dec 20 10:59:39 PST 2022
spring-boot-starter-webflux-version=2.5.12
swagger-core-version=1.6.2
spring-reactor-test-version=3.2.3.RELEASE
openapi-generator-version=5.4.0
einstein-bot-openapi-codegen-version=5.4.4
jackson-annotations-version=2.13.4
jackson-databind-nullable-version=0.2.1
maven-javadoc-plugin.version=3.2.0
base-sdk-package=com.salesforce.einsteinbot.sdk
maven-source-plugin.version=3.2.1
jackson-version=2.13.4.2
project.build.sourceEncoding=UTF-8
maven.compiler.source=1.8
spring-reactor-core-version=3.4.5
junit-jupiter-version=5.5.2
java.version=1.8.0_332
runtime-api-spec-url=https\://runtime-api-na-west.prod.chatbots.sfdc.sh/docs/specs/v5_0_0_api_specs.yml
maven.compiler.target=1.8
maven-release-plugin.version=3.0.0-M1
javax-annotation-version=1.3.2
api-spec-yaml-file=v5_0_0_api_specs.yml

```

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 8 above)
Comment (relango) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:293]: It's better not  to use the naming convention of file. Is it possible to read api-spec-yaml-file file in maven and use value of `info.version` ?

Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:293]: I could look into it, but also, what if we added a new maven property itself like `api-spec-yaml-file`? Something like
```
<api-spec-version>5.0.0</api-spec-version>
```
And then this property can be directly read in the code.

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 6 above)
Comment (atulkc) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:110]: Can you add more details to the exception? It should mention the version being currently used by the sdk and the versions that are supported. Anyone looking at the stack trace and exception message should be able to know exactly what they need to do

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 10]
```diff
@@ -246,4 +284,32 @@ private Mono<ClientResponse> mapErrorResponse(ClientResponse clientResponse) {
...
+    properties = new Properties();
+    properties.load(is);
+    return properties.getProperty("api-spec-version").replace("_", ".");
+  }
+
+  private String getLatestApiVersion() throws IOException {
+    SupportedVersions versions = getSupportedVersions();
+    Optional<SupportedVersionsVersions> supportedVersions = versions.getVersions()
+        .stream()
+        .filter(v -> Objects.equals(v.getStatus(), StatusEnum.ACTIVE))
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:301]: Can't you just do `v.getStatus() == StatusEnum.ACTIVE` ?

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 11]
```diff
@@ -86,10 +97,23 @@ void setHealthApi(HealthApi healthApi) {
...
   @Override
   public BotResponse startChatSession(RequestConfig config,
       ExternalSessionId sessionId,
       BotSendMessageRequest botSendMessageRequest) {
+    try {
+      if (!isApiVersionSupported()) {
+        String errorMessageFormat = "SDK failed to start chat as the API version is not supported. Current API version in SDK is %s, latest supported API version is %s, please upgrade to the latest version. Further information about specific versions is available here: https://github.com/forcedotcom/einstein-bot-sdk-java#version-numbering";
+        throw new RuntimeException(String.format(errorMessageFormat, getCurrentApiVersion(), getLatestApiVersion()));
+      }
+    } catch (IOException e) {
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:114]: Can you catch `IOException` in `getCurrentApiVersion` and rethrow as `RuntimeException` ? Also, let's create a separate exception like `UnsupportedSdkVersion` for this error.

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java [hunk 12]
```diff
@@ -246,4 +284,32 @@ private Mono<ClientResponse> mapErrorResponse(ClientResponse clientResponse) {
...
                 clientResponse.headers())));
   }
+
+  private String getCurrentApiVersion() throws IOException {
+    Properties properties;
+    InputStream is = getClass().getClassLoader()
+        .getResourceAsStream("properties-from-pom.properties");
+    properties = new Properties();
+    properties.load(is);
+    return properties.getProperty("api-spec-version").replace("_", ".");
```
Comment (relango) [inline src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:294]: Can you add a comment on why need to do this replacement?

File: src/test/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientTest.java [hunk 13]
```diff
@@ -329,4 +355,44 @@ public void testGetHealthStatus() {
...
+    ((BasicChatbotClientImpl) client).setVersionsApi(mockVersionsApi);
+  }
+
+  private void stubInactiveVersionsResponse() {
+    List<SupportedVersionsVersions> versions = new ArrayList<>();
+    SupportedVersionsVersions version = new SupportedVersionsVersions();
+    version.setVersionNumber("5.1.0");
+    version.setStatus(StatusEnum.ACTIVE);
+    versions.add(version);
+    SupportedVersions mockVersions = new SupportedVersions();
```
Comment (relango) [inline src/test/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientTest.java:392]: Can you refactor to avoid duplicate code in test methods?

**Review by relango — CHANGES_REQUESTED** [review]

File: src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java (hunk 11 above)
Comment (bharatsuri97) [reply src/main/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientImpl.java:114]: Makes sense, I will do that

File: src/test/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientTest.java (hunk 13 above)
Comment (bharatsuri97) [reply src/test/java/com/salesforce/einsteinbot/sdk/client/BasicChatbotClientTest.java:392]: Yes

**Review by relango — APPROVED** [review]

**Pull request merged**